### PR TITLE
ci(jenkins): codecov should happen once, for ruby:2.6

### PR DESCRIPTION
**Pull request by v1v**
## Highlights
- Matrix downstream job does run codecov for each run, let's run only once.

## Test Cases
- Without coverage

image

- With coverage

image

## Tasks
- [ ] Revert https://github.com/elastic/apm-agent-ruby/pull/485/files/d63bf53fa1056c84a80506fe3738f086fe544025#diff-8db37e8fcea0f1a8f2f39667e94ebcc4R208

### CODE REVIEW

**Comment by mikker** [comment]
Nice! Will this still report it for every version of Rails/Sinatra (the framework axis of the matrix)?

**Comment by v1v** [comment]
> Nice! Will this still report it for every version of Rails/Sinatra (the framework axis of the matrix)?

Indeed, it's not ideal, but the current implementation does care only for the ruby versions in the `.ci/Jenkinsfile` and the frameworks in the `.ci/downstreamTests.groovy`

I'm tempted to move that specific behavior to another file which contains the match of ruby-versions#frameworks that the coverage should be expected to, then `.ci/downstreamTests.groovy`
should read the file and ensure the matches.

So roughly something like:

```
cat .ci/.jenkins_coverage.yml
TUPLE:
  - ruby:2.6#rails-5.2
```

```
cat .ci/downstreamTests.groovy
...
     # read YAML file .ci/.jenkins_coverage.yml
     if (data[TUPLE].contains("${x}#") && data[TUPLE].contains("#${y}")) {
        steps.codecov(repo: "${steps.env.REPO}", basedir: "${steps.env.BASE_DIR}", secret: "${steps.env.CODECOV_SECRET}")
     }
...
```

What do you think?

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -197,19 +198,20 @@ def runBenchmark(version){
...
+def runJob(rubyVersion, boolean isCoverage=false){
   def branch = env.BRANCH_NAME
 
   if (env.CHANGE_ID) {
     branch = env.CHANGE_TARGET
   }
-  
-  build( job: "apm-agent-ruby/apm-agent-ruby-downstream/${branch}", 
+
+  build( job: "apm-agent-ruby/apm-agent-ruby-downstream/PR-485",
```
Comment (v1v) [inline .ci/Jenkinsfile]: This is only required for testing purposes as it does use the master branch as the target branch by default.

**Comment by mikker** [comment]
I think the extra file is a good idea 👍. One single source of truth.

File: .ci/.jenkins_codecov.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+# Tuple of ruby version and frameworks which are in charge to send data to codecov.
+# # is the separator between the ruby version and the framework.
+ENABLED:
+  - ruby:2.6#rails-5.2
```
Comment (mikker) [inline .ci/.jenkins_codecov.yml]: Sorry, Rails 6 came out in the meantime, so should probably be that instead 😆

Comment (v1v) [inline .ci/.jenkins_codecov.yml]:
```suggestion
  - ruby:2.6#rails-6.0.0
```

File: .ci/.jenkins_codecov.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+# Tuple of ruby version and frameworks which are in charge to send data to codecov.
+# # is the separator between the ruby version and the framework.
+ENABLED:
+  - ruby:2.6#rails-6.0.0
```
Comment (v1v) [inline .ci/.jenkins_codecov.yml]:
```suggestion
  - ruby:2.6#rails-6.0
```

File: .ci/Jenkinsfile (hunk 1 above)
Comment (v1v) [reply .ci/Jenkinsfile]:
```suggestion
  build( job: "apm-agent-ruby/apm-agent-ruby-downstream/${env.BRANCH_NAME}",
```

File: .ci/downstreamTests.groovy [hunk 4]
```diff
@@ -160,3 +162,14 @@ def runScript(Map params = [:]){
     }
   }
 }
+
+/**
+* Whether the given ruby version and framework are in charge of sending the
+* codecov results. It does require the workspace.
+*/
+def isCodecovEnabled(ruby, framework){
```
Comment (mdelapenya) [inline .ci/downstreamTests.groovy]: The `Enabled` part could be misleading compared to the javadocs, wdyt?

Comment (v1v) [reply .ci/downstreamTests.groovy]: Do you mean `isCodecov` rather than `isCodecovEnabled`?

**Comment by mikker** [comment]
💙💚💛💜❤️

**Pull request merged**